### ISSUE TITLE
feature request: system wide irods_environments.json defaults file

### ISSUE THREAD

**Issue by kript** [irods/irods#3783]
At the moment, (at least in 4.1.X, and I assume in 4.2.x) any user wishing to use icommands must have, or create, an `~/.irods/irods_environment.json` file with the required settings. More files (such as .irodsA) can be used, but are not required.

If there was a systems defaults file (say `/etc/default/` on Ubuntu), then everything in the irods_environments.json but the username could be specified there, and
the username could either be specified via environment variable, or the $USER used as a default, unless otherwise specified, with the irods_home making the same assumptions. If the `~/.irods/irods_environment.json` file exists, then those settings would be used in preference to the system defaults, and would allow for configurations where the iRODS user differed from the logged in user, for example..

This feature would mean that we would no longer have to create the environment files for users for them to use the system, which can be error prone, and hard to bulk edit later if a zone wide change is needed (this has happened twice to us in our iRODS history, so its not a theoretical issue).

Thanks to @mcast for the idea.

**Comment by trel** [comment]
Interesting...

The current behavior of the iCommands is:
1. If available, use `$IRODS_ENVIRONMENT_FILE` to define the location of an iRODS environment .json file
2. Then, if available, use `~/.irods/irods_environment.json` to define the iRODS environment
3. Else, require `iinit` to be run and prompt for the required values (and save those for future use by step 2.)

This proposal would be to inject a step between 2 and 3:
- 2.5. Then, if available, use a system-wide default file (I suggest located at `/var/lib/irods/clients/icommands/default_irods_environment.json` with permission `644` and part of the irods-icommands package) to define the iRODS environment.  Variable substitution would be performed for `$USER`.
- `ienv` would need to learn to report that it is using this default system configuration file.
- This is still 'all-in-one' - whichever environment file is being loaded/used needs to define ALL the required information.

OR

This proposal would create a kind of system-default mask that would be used to populated otherwise-undefined values coming from either step 1 or step 2 above.  This suggests a different ordering...
- 0. Load default values from `/var/lib/irods/clients/icommands/default_irods_environment.json` if available, perform variable substitution on `$USER`, if necessary.  Then, steps 1 and 2 would supplement/override these default values.
- `ienv` would need to differentiate and report the source of every value that it is using (some could be from one location, others could be coming from a different location).
- This is a more robust and flexible system, but a bigger change from the current file-based all-in-one ordering currently employed.

**Comment by kript** [comment]
Speaking for myself, I'd go for the latter 'mask'. I would probably push back on using `/var/lib/irods/clients/icommands/default_irods_environment.json` because;
1. as part of a package it may be overwritten on every upgrade, and take some time for config management to restore the previous state.
2. Its not one of the standard locations a system administrator would look for configuration files in (/etc, /opt, /usr/local/etc er etc..)

Instead, I'd like to make the case for checking for the existence of /etc/irods/irods_client_environment_defaults.json, which, _if it exists_, can set the defaults.

**Comment by trel** [comment]
Agreed on `/etc/irods/` as a better location - the `irods-icommands` package just didn't already have something there, but that's not a problem.

We would probably package the file as `*.template` (and document the feature/process of copying the *.template file) so the admin doesn't edit a packaged file.

Candidates:
1. `/etc/irods/irods_client_environment_defaults.json`
2. `/etc/irods/client_irods_environment_defaults.json`
3. `/etc/irods/default_irods_client_environment.json`
4. `/etc/irods/default_client_irods_environment.json`
5. `/etc/irods/irods_environment_defaults.json`
6. `/etc/irods/default_irods_environment.json`

I think I'm partial to 6..., as it keeps the existing eventual filename consistent (and only adds a leading adjective).

Another note is that we could implement this as 'all-in-one' at first, and then change the behavior later to be the 'mask' with ... minimal/no side-effects?  Might not be worth the effort, twice.

**Comment by kript** [comment]
Is the proposal is to have the package deploy a `/etc/irods/default_irods_environment.json.template` or `/var/lib/irods/clients/icommands/default_irods_environment.json.template` for the admin to copy to `/etc/irods/default_irods_environment.json`? I _think_ I prefer the latter, since its an opt-in feature.

I'm not sure of the rationale of starting with doing it one way and changing later - what does it provide over choosing the approach from the outset? If the idea was getting an `/etc/irods/default_irods_environment.json` file on the system first, then the 'mask' option allows for that as it would only source if the file existed..

**Comment by trel** [comment]
Yes, a template file is placed on the disk by the package.

An admin would decide to use the feature by copying and updating the template into the location that the code will look for.

The rationale for doing it one way and then changing -  would correspond with a 'small' change (a new place to check for an 'all-in-one') for 4-1-stable... and a 'bigger' change of existing behavior (mask) for 4-2-stable.    Alternately, we just work on the 'mask' for 4-2-stable?

**Comment by kript** [comment]
After some discussion at our end, we think we prefer you do the work once, so are voting for the mask for 4-2 stable.

Thank you!

**Comment by trel** [comment]
Sold.
